### PR TITLE
Add optional param to the identity client and api

### PR DESCRIPTION
**Pull request by kissmikijr**
## Hey, I just made a Pull Request!

continuation of: https://github.com/backstage/backstage/pull/14533

add optional parameter to the identity client and API

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/auth-node/src/DefaultIdentityClient.test.ts
Comment (punkle) [inline plugins/auth-node/src/DefaultIdentityClient.test.ts:420]: Should we assert that this has thrown an AuthenticationError so that the router knows what to return to the http response.

File: plugins/auth-node/src/DefaultIdentityClient.test.ts
Comment (punkle) [inline plugins/auth-node/src/DefaultIdentityClient.test.ts:426]: 👍

File: plugins/auth-node/src/DefaultIdentityClient.test.ts
Comment (punkle) [inline plugins/auth-node/src/DefaultIdentityClient.test.ts:404]: Im aware that I probably wrote this expectation. But I wonder is it possible to check the error type to make sure its an AuthenticationError?

File: plugins/playlist-backend/src/service/router.ts
Comment (punkle) [inline plugins/playlist-backend/src/service/router.ts:88]: This comment is pedantic and can be happily ignored.

Im wondering if we should decide on a name to call the response from this call and use it across the other plugins too.

We are already using: user, userIdentity, callerIdentity and identity.

File: plugins/auth-node/api-report.md
Comment (kissmikijr) [inline plugins/auth-node/api-report.md:50]: @Rugvip I tried the overload that you mentioned here https://github.com/backstage/backstage/pull/14533#discussion_r1040172973
Not sure I was successful, haha so please check if this is what you wanted to achieve with it.

File: plugins/permission-backend/src/service/router.test.ts
Comment (kissmikijr) [inline plugins/permission-backend/src/service/router.test.ts:90]: I added this any cast here because tsc wasn't passing because it assumed this function will be the other signature of the getIdentity method. Not sure how to do it properly , I wasn't really able to figure out how to mock an overloaded function in jest.

File: plugins/scaffolder-backend/src/service/router.ts
Comment (kissmikijr) [inline plugins/scaffolder-backend/src/service/router.ts:96]: I needed to convert this custom implementation of the IdentityApi to a class to be able to properly implement the overloaded function for it.

**Comment by kissmikijr** [comment]
yaay the build passed finally! :) @Rugvip I think this is ready for another round of reviews, this addresses most of the comments from this review https://github.com/backstage/backstage/pull/14533

**Comment by kissmikijr** [comment]
Not stale! Waiting for a review!

File: plugins/auth-node/api-report.md
Comment (Rugvip) [reply plugins/auth-node/api-report.md:50]: Not quite as I don't think this'd catch `optional: false`. Cleanest way is probably a type parameter for the request.

```ts
interface IdentityApiGetIdentityRequest<TOptional extends boolean = boolean> {
  optional: TOptional,
  ...
}

getIdentity(options: IdentityApiGetIdentityRequest<true>): Promise<BackstageIdentityResponse | undefined>;
getIdentity(options: IdentityApiGetIdentityRequest): Promise<BackstageIdentityResponse>;
```

**Review by Rugvip — CHANGES_REQUESTED** [review]
Quick comment but also wanted to check in on the urgency of this. Is this blocking something on your side, or just a bit more maintenance? Reason I'm asking is that I would prefer to take a more broader look at auth through #15999, but we're not there quite yet still

**Pull request closed without merging**
